app/models/concerns/taggable.rb

```diff
@@ -1,63 +1,23 @@
 module Taggable
-
-  def self.included(taggable)
-    taggable.class_eval do
-      attr_accessor :invalid_tags
-      attr_accessor :preview_mode, :placeholder_tags
-
-      has_many :filter_taggings, as: :filterable
-      has_many :filters, through: :filter_taggings
-      has_many :direct_filter_taggings, -> { where(inherited: 0) }, class_name: "FilterTagging", as: :filterable
-      has_many :direct_filters, source: :filter, through: :direct_filter_taggings
-
-      has_many :taggings, as: :taggable, dependent: :destroy
-      has_many :tags, through: :taggings, source: :tagger, source_type: 'Tag'
-
-      has_many :ratings,
-        -> { where("tags.type = 'Rating'") },
-        through: :taggings,
-        source: :tagger,
-        source_type: 'Tag',
-        before_remove: :remove_filter_tagging
-      has_many :categories,
-        -> { where("tags.type = 'Category'") },
-        through: :taggings,
-        source: :tagger,
-        source_type: 'Tag',
-        before_remove: :remove_filter_tagging
-      has_many :archive_warnings,
-        -> { where("tags.type = 'ArchiveWarning'") },
-        through: :taggings,
-        source: :tagger,
-        source_type: 'Tag',
-        before_remove: :remove_filter_tagging
-      has_many :fandoms,
-        -> { where("tags.type = 'Fandom'") },
-        through: :taggings,
-        source: :tagger,
-        source_type: 'Tag',
-        before_remove: :remove_filter_tagging
-      has_many :relationships,
-        -> { where("tags.type = 'Relationship'") },
-        through: :taggings,
-        source: :tagger,
-        source_type: 'Tag',
-        before_remove: :remove_filter_tagging
-      has_many :characters,
-        -> { where("tags.type = 'Character'") },
-        through: :taggings,
-        source: :tagger,
-        source_type: 'Tag',
-        before_remove: :remove_filter_tagging
-      has_many :freeforms,
-        -> { where("tags.type = 'Freeform'") },
-        through: :taggings,
-        source: :tagger,
-        source_type: 'Tag',
-        before_remove: :remove_filter_tagging
-
-      after_update :reset_placeholders
+  extend ActiveSupport::Concern
+
+  included do
+    attr_accessor :invalid_tags
+    attr_accessor :preview_mode, :placeholder_tags
+
+    has_many :taggings, as: :taggable, inverse_of: :taggable, dependent: :destroy
+    has_many :tags, through: :taggings, source: :tagger, source_type: 'Tag'
+
+    Tag::VISIBLE.each do |type|
+      has_many type.underscore.pluralize.to_sym,
+               -> { where(tags: { type: type }) },
+               through: :taggings,
+               source: :tagger,
+               source_type: 'Tag',
+               before_remove: :destroy_tagging
```

Comment:
They're automatically deleted when the tag is removed from the work. But the automatic deletion uses `delete` instead of `destroy` on the taggings, so it was impossible to use callbacks on the taggings before.

Reply:
:+1:, thanks!